src/sorting/quicksort.jl

```diff
@@ -0,0 +1,416 @@
+        partition = consolidate_batch_partition(vals, pivot, lo, L, N_t, N_b, i, global_b_sums)
+    end
+
+    #= step 4: recursion. If the first subproblem doesn't require global mem,
+    run it on a temporary stream. Otherwise run in default stream, so that
+    the two subproblems run serially, and don't conflict over the global array=#
+    if i == 1 && partition > lo
+        if partition - lo <= N_t ^ 2
+            s = CuDeviceStream()
+            @cuda threads=N_t dynamic=true stream=s qsort_kernel(vals, lo, Int32(partition), global_b_sums, block_dim, Val(!P), depth + Int8(1))
+            CUDA.unsafe_destroy!(s)
+        else
+            @cuda threads=N_t dynamic=true qsort_kernel(vals, lo, Int32(partition), global_b_sums, block_dim, Val(!P), depth + Int8(1))
+        end
+    end
+
+    if i == 1 && partition < hi
+        if hi - partition <= N_t ^ 2
+            s = CuDeviceStream()
+            @cuda threads=N_t dynamic=true stream=s qsort_kernel(vals, Int32(partition), hi, global_b_sums, block_dim, Val(!P), depth + Int8(1))
+            CUDA.unsafe_destroy!(s)
+        else
+            @cuda threads=N_t dynamic=true qsort_kernel(vals, Int32(partition), hi, global_b_sums, block_dim, Val(!P), depth + Int8(1))
+        end
+    end
+    return nothing
+end
+
+function quicksort!(c :: CuArray{T}, block_size :: Integer) where T
+    @assert isinteger(Base.log2(block_size)) "`block_size` must be a power of two"
+    CUDA.limit!(CUDA.LIMIT_DEV_RUNTIME_SYNC_DEPTH, 24)
+    N = length(c)
+    b_sums = CuArray{Int32}(undef, ceil_div(N, block_size))
+    CUDA.@sync(@cuda threads=block_size qsort_kernel(c, Int32(0), N, b_sums, Val(Int32(block_size)), Val(true), Int8(1)))
+end
+
+"""
+Attempts to determine the maximum blocksize for running quicksort on the
+available hardware. Picks the greatest power of two which doesn't violate:
+    - Inherent maximum block size
+    - Register usage
+    - Shared memory
+This assumes that register usage doubles with elevated debug level, but usage of
+other resources is unaffected.
+"""
+function quicksort_block_size(T)
+    # requirements
+    # this is a guess for older architectures. 64 is fine for RTX 2070
+    registers_per_thread = 128
+    if Base.JLOptions().debug_level >= 2
+        registers_per_thread *= 2
+    end
+    shmem_per_thread = sizeof(Int32) + sizeof(T)
+
+    # resource caps
+    MAX_BLOCK = CUDA.attribute(device(), CUDA.DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_X)
+    MAX_REG = CUDA.attribute(device(), CUDA.DEVICE_ATTRIBUTE_MAX_REGISTERS_PER_BLOCK)
+    MAX_SHMEM = CUDA.attribute(device(), CUDA.DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK)
+
+    # this is safe assuming a 1d block is always executed on 1 multiprocessor
+    block_size = 1
+    while all(block_size .* (1, registers_per_thread, shmem_per_thread) .< (MAX_BLOCK, MAX_REG, MAX_SHMEM))
+        block_size *= 2
+    end
+    return block_size
+end
+
+function Base.sort!(c :: CuArray{T}) where T
```

Comment:
`Base.sort!` has quite some kwargs, can some of those be implemented already by your `quicksort!` method?

Reply:
Some certainly can. `dims` is easy by using views. `rev` is possible with minor modifications. These two I plan to include in this PR.

Others are harder. `by` could be done in two ways:
1. Custom compiled kernels.
2. Broadcasting `by` to an array which would be our keys, and then sort has to move twice as much data (keys and values).
Of the two, I prefer 1. Either way, this probably has to be a follow-up feature.

`lt` could be defined similarly, but it could only be done if there is always a corresponding `==`. This is because the sort alternates between partitioning on `<` and `<=`.

The other two args, `alg` and `order` do not seem applicable. I can't even find a good example of what `order` does.